org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corext/template/java/CodeSnippetTemplates.java

```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.corext.template.java;
+
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jface.text.templates.Template;
+
+public enum CodeSnippetTemplates {
+
+	SYSTRACE(TemplatePreferences.SYSTRACE_ID, JavaContextType.ID_STATEMENTS, TemplatePreferences.SYSTRACE_CONTENT, TemplatePreferences.SYSTRACE_DESCRIPTION),
+	FOREACH(TemplatePreferences.FOREACH_ID, JavaContextType.ID_STATEMENTS, TemplatePreferences.FOREACH_CONTENT, TemplatePreferences.FOREACH_DESCRIPTION),
+	FORI(TemplatePreferences.FORI_ID, JavaContextType.ID_STATEMENTS, TemplatePreferences.FORI_CONTENT, TemplatePreferences.FORI_DESCRIPTION);
+
+	private final String templateId;
+	private final String contextType;
+	private final String defaultContent;
+	private final String description;
+
+	private CodeSnippetTemplates(String templatesId, String contextType, String defaultContent, String description) {
+		this.templateId = templatesId;
+		this.contextType = contextType;
+		this.defaultContent = defaultContent;
+		this.description = description;
+	}
+
+	public Template createTemplate() {
+		return new Template(this.name().toLowerCase(), this.description, this.contextType, this.defaultContent, false);
+	}
+
+	public String getId() {
+		return this.templateId;
+	}
+}
+
+class TemplatePreferences {
+	// IDs
+	public static final String SYSTRACE_ID = "org.eclipse.jdt.ui.templates.systrace";
```

Comment:
do we need to keep references to jdt.ui?

Reply:
I'm okay to rename the id field. How about `org.eclipse.jdt.ls.templates.XXXX`?

Reply:
+1